### ISSUE TITLE
AwaitableUnaryClientCallListener still holds lock after onMessage

### ISSUE THREAD

**Issue by meltsufin** [openzipkin/zipkin-gcp#100]
Contrary to the javadoc on `AwaitableUnaryClientCallListener` that says `Blocks until {@link #onMessage} or {@link #onClose}.`, it continues blocking after `onMessage` until `onClose` is called.

Potentially related to: https://github.com/spring-cloud/spring-cloud-gcp/issues/1118 where we're seeing hanging threads.

**Comment by adriancole** [comment]
agreed. the next task is to figure out why awaiting until a status was received is a problem

**Comment by adriancole** [comment]
Thanks for reporting the doc bug. Feel free to open another issue if there's a new behavior we need to be aware of, or if there is a condition which escapes the onClose callback.

**Comment by meltsufin** [comment]
Thanks for fixing the doc. What do you think of having a timeout when waiting on the count-down-latch?

**Comment by adriancole** [comment]
well, it sounds like working around a state bug :P

**Comment by adriancole** [comment]
I guess what I mean to say is if there is a state bug, then if we add a
timeout, whereever the exposure to that bug is would be possible to hit
that timeout on each request. In your setup can you verify that netty, etc
is aligned properly? Wondering where the cases are that the finish hook
might not be called as that seems something that would affect others...
perhaps there's a missing try/finally somewhere?

>

**Comment by meltsufin** [comment]
You can see some results of the investigation here: https://github.com/spring-cloud/spring-cloud-gcp/issues/1118.
I'm still not sure why it would ever be better to wait indefinitely than some specified timeout. Waiting indefinitely just masks the problem longer, whereas a failure at timeout at least gives an earlier indication that something is wrong.

**Comment by adriancole** [comment]
If you want to raise a pull request, trap the timeout exception and throw
as an IllegalStateException saying something like possible bug, that could
make sense cc @anuraaga
